### ISSUE TITLE
Why yarn? Any chance pnpm could eventually be supported (in the name of saving disk space)?

### ISSUE THREAD

**Issue by matronator** [redwoodjs/graphql#2907]
I looked (okay, not that hard, but I definitely didn't see it mentioned in the tutorial), but couldn't find an explanation of why is yarn a requirement? Can you please clarify to me, why yarn?

I primarily use [pnpm](https://pnpm.io) as my package manager, mostly because I'm slightly worried about my 8 years old MacBook's internal SSD running out of lifecycles due to excessive `npm install` garbage. And yarn isn't much better in this, it installs just as much garbage as npm. So anyways, onto my second question:

Is there any possibility in the future, that npm/pnpm would be supported as well and not just yarn?

Thank you!

**Comment by peterp** [comment]
@matronator Yup, there is no explicit reason, other than we have some technical debt that we've been cleaning up. We're getting there!

@dac09 Could we run an experiment on the RedwoodJS project side to see if it's possible?

**Comment by thedavidprice** [comment]
We use Yarn for Yarn Workspaces. NPM did not support Workspaces until recently.

General FYI: I did an experiment two weeks ago using Yarn 2 with the Redwood Framework — even without pnpm there were major breaking changes and substantial behavior changes to dependency management and resolutions. Importantly, they are handling Workspaces differently in 2. Nothing a dealbreaker, but it did make me inclined to wait for Yarn 3 GA before re-evalutating.

**Comment by ajcwebdev** [comment]
If we decide to move forward with Yarn 3 we should take a look at what a migration to `pnpm` or any other package manager would look like and how it would compare to upgrading. There's a lot of innovation happening right now in this space so it'll be worth taking some time to make sure we're future proofing ourselves as much as possible.

**Comment by jtoar** [comment]
StackBlitz also teased their own package manager; they haven't released it yet (I'm assuming they will one day)—just another one to consider.

**Comment by thedavidprice** [comment]
Early days, but NPM v7 might be a case similar to Yarn v2 in regards to some big, fundamental changes. Hot take. Needs verification.

**Comment by Albert-Gao** [comment]
> Early days, but NPM v7 might be a case similar to Yarn v2 in regards to some big, fundamental changes. Hot take. Needs verification.

Correct, I ditched npm v7 and started using Yarn 1 because of the breaking change in v7, just do not want to use that --legacy-peer-dep arg.  😂

**Comment by ajcwebdev** [comment]
## [Yarn 3 announcement](https://dev.to/arcanis/yarn-3-0-performances-esbuild-better-patches-e07)

### Grist for the mill

>While the migration from Yarn 1 to Yarn 2 brought some discomfort, the migration to Yarn 3 should prove easier - regardless of the version you come from. The user-facing breaking changes we made this season are mostly little details that may only affect you in very specific cases:
>
>* Node 10 isn't supported anymore
>* Plug'n'Play hooks are now called `.pnp.cjs` (vs `.pnp.js`)
>* Virtual folders are now called `__virtual__` (vs `$virtual`)
>* The [editor SDKs](https://yarnpkg.com/getting-started/editor-sdks) have been moved to `@yarnpkg/sdks`
>* Etc; full list [here](https://github.com/yarnpkg/berry/blob/master/CHANGELOG.md)
>
>Even for Yarn 1 users, migrating from 1 to 3 should be easier: we made it so that Yarn will detect when this situation arises to then automatically enable the `node-modules` linker. That alone should address most of the problems you may have been hitting when attempting the upgrade - and for everything else, make sure to take a look at our [Migration Guide](https://yarnpkg.com/getting-started/migration#step-by-step) which got significantly improved over the past year.

**Comment by ajcwebdev** [comment]
From [Maël](https://twitter.com/arcanis/status/1419720778375913473), lead Yarn maintainer:

>*I'd recommend to attempt the migration and ping us on Discord once you hit a blocker - if you use the n_m linker it should be straightforward, although I don't know what was "workspaces [are handled] differently in 2" was referring to 🤔*

**Comment by thedavidprice** [comment]
Interesting. I'm curious about taking this for a test run. Although I have strong doubt about this quote in particular --> "Even for Yarn 1 users, migrating from 1 to 3 should be easier". Lerna and Yarn Workspaces got into a fight when I tried to upgrade to v2 for a POC. 

Overall, this feels like good news. I just think it's going to take about 6 months to see where the ecosystem lands in terms of support. E.g. even if we upgrade to Yarn v3 for the framework, will projects using v3 have build support on Netlify, Vercel, Render, etc...

**Comment by peterp** [comment]
I'm trying to get to the point where we don't dictate the package manager for Redwood projects. 

My goal is that we create single binary that encases the build and design time packages. The user will only install the run-time packages: express, react, apollo-graphql, etc.

Then we don't have to rely on workspaces and a user is free to use whatever they prefer as a package manager.

A bunch of added advantages: crazy fast installations.

**Comment by larixer** [comment]
@thedavidprice 

> Overall, this feels like good news. I just think it's going to take about 6 months to see where the ecosystem lands in terms of support. E.g. even if we upgrade to Yarn v3 for the framework, will projects using v3 have build support on Netlify, Vercel, Render, etc...

You can use `node_modules` installs with Yarn 2 and Yarn 3. With Yarn 3 you can also benefit from content-addressable storage (`nmMode: hardlinks-global`), when all your unique files from different packages will be shared across multiple projects in the same place on disk - the same idea as used by `pnpm` and requested by @matronator to save disk space, but without modifications to node_modules layout via symlinks, hence it should be compatible to all existing libraries for `npm`.  

You don't need to wait for ecosystem to be ready for Yarn 3 if you will continue to use `node_modules`.

**Comment by thedavidprice** [comment]
Hi @larixer! Thank you for jumping in here.

These all sound like great benefits. Definitely want to look more into understanding `nmMode: hardlinks-global` — sounds very nice! And know that all along I've been excited and impressed with the heavy lifting the Yarn team has been doing. We literally couldn't have built Redwood the way we did without Yarn.

I did use node_modules in my attempt to upgrade to Yarn v2 in our Framework monorep. What I discovered was a lot of emerging resolution issues, likely due to improvements that revealed issues we've had along. More importantly, our use and setup of Yarn Workspaces threw a lot of errors and broke the scripts/tools we have in place. If you're curious (and have time for some suggestions), take a look at our scripts in the project's root package.json here:
https://github.com/redwoodjs/redwood/blob/d9d385244ec3a5a6d6b8da5c58dc3295676f54e2/package.json#L24-L38

Overall, it felt like there was a path forward. Just not a clean one.

But know I'm motivated to give Yarn v3 a trial run in the near future!

**Comment by larixer** [comment]
> What I discovered was a lot of emerging resolution issues, likely due to improvements that revealed issues we've had along. More importantly, our use and setup of Yarn Workspaces threw a lot of errors and broke the scripts/tools we have in place. If you're curious (and have time for some suggestions), take a look at our scripts in the project's root package.json here:
> https://github.com/redwoodjs/redwood/blob/d9d385244ec3a5a6d6b8da5c58dc3295676f54e2/package.json#L24-L38

Yes, I'm curious and willing to take a look. I am not sure what I'm looking at, you referenced one commit and I don't understand what exactly have you tried to do and what problems have you faced? Do you have some Pull Request where you have tried to migrate to Yarn 2 with description of problems you faced? Alternatively are you comfortable to pop on our discord: 
https://discord.com/invite/yarnpkg
where we can discuss the issues you faced in real time?

**Comment by thedavidprice** [comment]
Well, look at that "Upgrade Framework to Yarn v3" #3154 

😉

**Comment by thedavidprice** [comment]
@matronator thank you for trying this out and reporting back here! I think your experience verifies what we've been assuming is the next steps on our end:
- resolve Yarn warnings across the RW framework packages
- resolve CLI bin requirements

I'm not sure if there's a path forward directly from a project, unfortunately.

**Comment by mwiemer-microsoft** [comment]
Also here because of Bun 1.0, would love to finally escape the slow installs of yarn and npm, ref https://github.com/oven-sh/bun/issues/4207 and [https://bun.sh](Bun - a fast all-in-one JavaScript toolkit) for more info, integration seems blocked on changes here and it's been nearly two years since a substantial update :(

**Comment by thedavidprice** [comment]
@guilt @mwiemer-microsoft 

When Bun, as a runtime, can support full node compatibility and, for example, work with Fastify and Pino, it would be very interesting to explore running Redwood on Bun. As of this time, I'm unsure how Bun could even be used to run Redwood's dev server (Fastify and Vite currently).

There's a lot to like about Bun. And I hope it pushed the ecosystem in the right direction. But given that Redwood already solves, out of the box, the challenges of toolchain integration, I'd be curious to better understand what it is that is currently of high enough value to make a swap? And, additionally, if you're aware of what those tradeoffs might be?

If you're about to answer "performance", please do take a look at this excellent writeup by @mcollina:
https://buff.ly/3r9Vhm0

Here's an important excerpt:
> In fact, pnpm --prefer-offline flag has roughly the same behavior of bun install and it yields performance in the same ballpark...

And, it also just so happens that Yarn 3 is as performant as pnpm (I know, right? Who knew 🤷‍♂️):
[Package Manager Benchmarks](https://p.datadoghq.eu/sb/d2wdprp9uki7gfks-c562c42f4dfd0ade4885690fa719c818?refresh_mode=sliding&tpl_var_npm%5B0%5D=%2A&tpl_var_pnpm%5B0%5D=%2A&tpl_var_yarn-classic%5B0%5D=%2A&tpl_var_yarn-modern%5B0%5D=%2A&tpl_var_yarn-nm%5B0%5D=%2A&tpl_var_yarn-pnpm%5B0%5D=no&from_ts=1686802631906&to_ts=1694578631906&live=true)

I know Node performance has been lagging, but I'm very optimistic about Node 20. And I'm very hopeful that Bun will further push Node's overall performance roadmap:
https://blog.rafaelgss.dev/state-of-nodejs-performance-2023

---

## 🚢 Shippers Gotta Ship
Lastly, I get that Redwood simply might not support the tools you want to use. Or, frankly, support you easily swapping out tools you use when you want to. There are a lot of things we could to better to this end.

But we choose not to. For one reason, we build with Redwood and we've chosen to accept good enough.

For a more important reason, we're investing our limited time and resouces in shipping innovative new features instead of futzing around with tooling. ('Cause, frankly, it's a lot more interesting and we don't really like futzing around with tooling even though we signed up for it as framework authors.)

### New Features: Redwood Realtime and Mailer + Studio
If you haven't checked out [v6.2](https://github.com/redwoodjs/redwood/releases/tag/v6.2.0) yet, I suggest you do because I hope you'll get excited about [Redwood Realtime](https://youtu.be/ChjhR89nMyg?t=350) and [Redwood Mailer API + Studio](https://youtu.be/ChjhR89nMyg?t=1821) integration.

### React SSR, Streaming, and Server Components
We have an experimental version of [Redwood SSR + Streaming](https://community.redwoodjs.com/t/react-streaming-and-server-side-rendering-ssr/5052/) that's availble. And we created a very fun production app that runs on Fly.io, using Neon.tech Postgres, with Resend.com for emails. You can [learn more about that here](https://twitter.com/selfteachme/status/1701378860342763566?s=20).

Lastly, we're well into being the second major framework to suport RSC, which you can [try as an experimental feature today](https://community.redwoodjs.com/t/react-server-components-rsc/5081/7).

These are the things that are important to us, along with the other items on our Bighorn Epoch roadmap: redwoodjs.com/roadmap

This priorities might not resonate with you and your requirements. I totally get that. But if they do resonate, Redwood is an open source project for a reason — we welcome and would love your help! 🤗

**Comment by mwiemer-microsoft** [comment]
Thanks @thedavidprice for the thorough response, anything is better than nothing when it comes to longstanding open issues. I'll admit I'm no expert here--I got through half the Redwood tutorial when it was v2, that was 6+ months ago. I was impressed. Bun obviously has some catchy graphs advertising speed and a low-level language behind it that makes those claims almost believable, but I haven't built anything with it yet. I was asking more out of curiosity and hope then anything grounded.

Thanks for clarifying your position here, it makes a lot of sense that the goal of Redwood is basically to be "Bun + backend," and I think you guys are doing a great job. No need to support Bun, really, just wanted an update on this issue in general, and you provided that 😄

**Comment by AddictArts** [comment]
FWIW: Yarn is a full stop no go, can't use it. I am passing on Redwood due to inflexibility in allowing the user to decide on what to use.

**Comment by peterp** [comment]
@AddictArts Absolutely hear you. The ability to pick a package manager is something that we'll make happen.

**Comment by toxsick** [comment]
@peterp what has changed in regard to your previous comment? Still hope pnpm support will happen at some point.

**Comment by peterp** [comment]
@toxsick Nothing! We are busy with "the unbundling" and this will make it a reality.
